File: expo-app/targets/widgets/WidgetData.swift
import Foundation

/// Die Nutzlast, wie sie die Web-App über die Brücke schickt.
///
/// SIE ENTHÄLT TEXT, KEINE DATEN – und das ist Absicht. Ein Widget hat
/// keinen Zugriff auf die Übersetzungen der Web-App und auf nichts aus
/// `shared/`. Bekäme es Zahlen und Datumsangaben, müsste hier die halbe
/// App noch einmal entstehen: vier Sprachen, Pluralformen, Datumsformate.
/// Stattdessen rechnet und formuliert die Web-App fertig, und dieser Teil
/// malt nur. Die Gegenseite steht in `shared/widgetData.ts`; wer hier ein
/// Feld ändert, muss dort ebenfalls hin.
struct WidgetPanel: Codable {
  /// Blickfang – Zahl oder kurzes Zeichen («3», «✓», «–»).
  let value: String
  /// Worum es geht («Seeblick», «Kühlbox»).
  let title: String
  /// Ergänzung in klein («noch 3 Tage»).
  let subtitle: String
  /// Ziel beim Antippen – ein Pfad der App («/tagebuch/7»).
  let url: String
}

/// Fortschritts-Ring, 0–1.
struct WidgetProgress: Codable {
  let ratio: Double
  let label: String
}

/// Eine Zeile der Liste zum Abhaken (#327).
///
/// `kind` unterscheidet Packliste und Ämtli. Beide zählen ihre Nummern
/// getrennt – ohne die Unterscheidung würde ein Häkchen auf Packlisten-
/// Eintrag 7 unterwegs zum Ämtli 7 werden.
struct WidgetTask: Codable, Identifiable {
  let id: Int
  let kind: String
  let title: String
  let checked: Bool
}

struct WidgetPayload: Codable {
  let version: Int
  let builtAt: String
  let trip: WidgetPanel
  let packing: WidgetProgress?
  let supplies: WidgetPanel
  let tasksTitle: String
  let tasksEmpty: String
  let tasksUrl: String
  let tasks: [WidgetTask]
}

/// Ein im Widget gesetztes Häkchen, das noch nicht beim Server ist (#327).
///
/// Die Erweiterung hat die Sitzung der App nicht und kann den Server
/// weder fragen noch ihm etwas sagen. Sie legt das Häkchen deshalb im
/// gemeinsamen Ordner ab; die App liest es beim nächsten Start aus und
/// schiebt es in ihre bestehende Warteschlange. Die Gegenstelle steht in
/// `shared/widgetActions.ts` – wer hier ein Feld ändert, muss dort hin.
struct PendingAction: Codable {
  let kind: String
  let itemId: Int
  let checked: Bool
  /// Millisekunden seit 1970 – entscheidet, welcher Wert gewinnt.
  let at: Double
}

extension WidgetPayload {
  /// Fassung, die dieser Widget-Stand versteht.
  static let supportedVersion = 2

  /// Was angezeigt wird, solange die App noch nie gelaufen ist.
  ///
  /// EIN WIDGET IST NIE LEER: Es wird in der Galerie als Vorschau
  /// gezeigt, bevor überhaupt Daten existieren, und es steht auf dem
  /// Bildschirm, wenn jemand die App nach dem Neustart des Geräts noch
  /// nicht geöffnet hat. Ein weisses Rechteck sähe dann nach einem
  /// Fehler aus. Die Texte sind deutsch, weil hier keine Sprache bekannt
  /// ist – sobald die App einmal offen war, kommen sie in der
  /// eingestellten Sprache von dort.
  static let placeholder = WidgetPayload(
    version: supportedVersion,
    builtAt: "",
    trip: WidgetPanel(
      value: "–",
      title: "CampMesser",
      subtitle: "App öffnen für aktuelle Daten",
      url: "/"
    ),
    packing: nil,
    supplies: WidgetPanel(
      value: "–",
      title: "Vorräte",
      subtitle: "App öffnen für aktuelle Daten",
      url: "/kuehlbox"
    ),
    tasksTitle: "Zum Abhaken",
    tasksEmpty: "App öffnen für aktuelle Daten",
    tasksUrl: "/packlisten",
    tasks: []
  )
}

/// Der gemeinsame Ordner von App und Widget.
///
/// Die Kennung MUSS zu `app.json` passen (`ios.entitlements`). Stimmen
/// die beiden nicht überein, schreibt die App in einen Ordner, den das
/// Widget nicht liest – ohne Fehlermeldung, das Widget zeigt einfach
/// ewig den Platzhalter. Deshalb steht sie hier an genau einer Stelle.
enum WidgetStore {
  static let appGroup = "group.ch.campmesser.app"
  static let key = "campmesserWidget"
  /// Zweiter Schlüssel: die im Widget gesetzten Häkchen (#327). Er muss
  /// zu `WIDGET_ACTIONS_KEY` in `expo-app/App.js` passen.
  static let actionsKey = "campmesserWidgetActions"
  /// Wie `PENDING_LIMIT` in `shared/widgetActions.ts`.
  static let pendingLimit = 100

  private static var defaults: UserDefaults? {
    UserDefaults(suiteName: appGroup)
  }

  /// Die zuletzt geschriebene Nutzlast lesen.
  ///
  /// Bei allem, was nicht passt – kein Eintrag, kaputtes JSON, eine
  /// Fassung aus der Zukunft – kommt der Platzhalter zurück. Ein Widget
  /// darf nie abstürzen: Es läuft im Auftrag des Systems, und ein
  /// Absturz kostet das Aktualisierungs-Budget der ganzen App.
  ///
  /// ÜBER DEN STAND AUS DER APP LEGT SICH, WAS IM WIDGET GETIPPT WURDE:
  /// Die App weiss davon erst beim nächsten Start. Ohne diese Überlagerung
  /// spränge der Schalter zurück, sobald das Widget neu zeichnet – und
  /// nichts wirkt kaputter als ein Schalter, der nicht bleibt.
  static func load() -> WidgetPayload {
    guard
      let raw = defaults?.string(forKey: key),
      let data = raw.data(using: .utf8),
      let payload = try? JSONDecoder().decode(WidgetPayload.self, from: data),
      payload.version == WidgetPayload.supportedVersion
    else {
      return .placeholder
    }
    return payload.withTasks(merge(payload.tasks, pending()))
  }

  /// Die vorgemerkten Häkchen lesen (Unsinn zählt als leer).
  static func pending() -> [PendingAction] {
    guard
      let raw = defaults?.string(forKey: actionsKey),
      let data = raw.data(using: .utf8),
      let queue = try? JSONDecoder().decode([PendingAction].self, from: data)
    else {
      return []
    }
    return queue
  }

  /// Ein im Widget gesetztes Häkchen vormerken.
  ///
  /// Mehrfaches Umschalten desselben Eintrags ergibt EINEN Eintrag mit dem
  /// zuletzt gewählten Wert – genau wie `addPending` in
  /// `shared/widgetActions.ts`.
  static func remember(kind: String, itemId: Int, checked: Bool) {
    let target = pendingKey(kind, itemId)
    var queue = pending().filter { pendingKey($0.kind, $0.itemId) != target }
    queue.append(
      PendingAction(
        kind: kind,
        itemId: itemId,
        checked: checked,
        at: Date().timeIntervalSince1970 * 1000
      )
    )
    if queue.count > pendingLimit {
      queue.removeFirst(queue.count - pendingLimit)
    }
    guard
      let data = try? JSONEncoder().encode(queue),
      let text = String(data: data, encoding: .utf8)
    else { return }
    defaults?.set(text, forKey: actionsKey)
  }

  /// Gegenstelle zu `mergePending`: Bei mehreren Einträgen zum selben Ziel
  /// gewinnt der jüngste – deshalb nach `at` sortiert und nicht nach
  /// Reihenfolge im Speicher.
  static func merge(_ tasks: [WidgetTask], _ pending: [PendingAction])
    -> [WidgetTask]
  {
    guard !pending.isEmpty else { return tasks }
    var latest: [String: PendingAction] = [:]
    for entry in pending.sorted(by: { $0.at < $1.at }) {
      latest[pendingKey(entry.kind, entry.itemId)] = entry
    }
    return tasks.map { task in
      guard let hit = latest[pendingKey(task.kind, task.id)] else { return task }
      return WidgetTask(
        id: task.id,
        kind: task.kind,
        title: task.title,
        checked: hit.checked
      )
    }
  }

  /// Eine Zeile pro Listeneintrag – Packliste und Ämtli zählen getrennt.
  private static func pendingKey(_ kind: String, _ itemId: Int) -> String {
    "\(kind):\(itemId)"
  }
}

extension WidgetPayload {
  /// Dieselbe Nutzlast mit ausgetauschter Liste.
  func withTasks(_ replacement: [WidgetTask]) -> WidgetPayload {
    WidgetPayload(
      version: version,
      builtAt: builtAt,
      trip: trip,
      packing: packing,
      supplies: supplies,
      tasksTitle: tasksTitle,
      tasksEmpty: tasksEmpty,
      tasksUrl: tasksUrl,
      tasks: replacement
    )
  }
}

/// Adresse, die beim Antippen geöffnet wird.
///
/// `campmesser://öffnen?pfad=/tagebuch/7` – das Schema steht in
/// `app.json`, die Gegenstelle in `expo-app/App.js`. Nur Pfade mit
/// führendem «/» werden gebaut; alles andere führt auf die Startseite.
func widgetURL(_ path: String) -> URL? {
  guard path.hasPrefix("/"), !path.hasPrefix("//") else {
    return URL(string: "campmesser://open?path=/")
  }
  var components = URLComponents()
  components.scheme = "campmesser"
  components.host = "open"
  components.queryItems = [URLQueryItem(name: "path", value: path)]
  return components.url
}

File: shared/widgetData.ts
import { l4, pick, type Language } from "./i18n";
import { daysUntilTrip, currentTripDay } from "./trips";
import { expiryInfo } from "./food";
import { gearTaskDue, type GearTaskLike } from "./gearTasks";
import type { WidgetTask } from "./widgetActions";
import { tripDisplayName } from "./tripName";

/**
 * Die Nutzlast für die iPhone-Widgets (#323).
 *
 * DAS WIDGET BEKOMMT TEXT, KEINE DATEN. Das ist die zentrale Entscheidung
 * hier, und sie hat einen handfesten Grund: Ein WidgetKit-Widget ist Swift
 * und lebt in einer eigenen Erweiterung – es hat keinen Zugriff auf die
 * Übersetzungen der Web-App, nicht auf `Intl`, nicht auf die Logik in
 * `shared/`. Würde es Zahlen und Datumsangaben bekommen, müsste die halbe
 * App in Swift nachgebaut werden: vier Sprachen, Pluralformen,
 * Datumsformate – und beim nächsten Textwechsel driften die beiden
 * Fassungen auseinander, ohne dass es jemand merkt.
 *
 * Deshalb rechnet und formuliert die Web-App fertig, und das Widget malt
 * nur noch. Es kennt keine Reise, keinen Zeitplan und keine Sprache – es
 * kennt eine Überschrift, eine Zeile darunter und eine Zahl für den Ring.
 *
 * WAS DAS KOSTET: Nach einem Sprachwechsel stehen die alten Texte im
 * Widget, bis die App das nächste Mal offen war. Das ist verkraftbar –
 * die Alternative wäre eine zweite, stillschweigend veraltende
 * Übersetzung im Swift-Teil.
 *
 * WAS DRIN STEHT, ist bewusst knapp: Ein Widget wird im Vorbeigehen
 * angeschaut. Zwei Fragen beantwortet es, mehr nicht – «wann geht es los
 * und bin ich bereit» und «muss ich etwas aufbrauchen oder pflegen».
 *
 * DAZU EINE LISTE ZUM ABHAKEN (#327): Seit iOS 17 kann ein Widget etwas
 * TUN. Was dort steht, hängt davon ab, wo man gerade steht – vor der
 * Reise die nächsten offenen Punkte der Packliste, während der Reise die
 * heutigen Ämtli. Beides sind Handgriffe, die man im Vorbeigehen macht,
 * und beide sind idempotent: «Eintrag 42 ist abgehakt» schadet doppelt
 * geschickt nicht.
 */

/** Ein Widget-Feld: grosse Zahl, Titel, Zeile darunter. */
export interface WidgetPanel {
  /** Blickfang – Zahl oder kurzes Zeichen («3», «heute», «–»). */
  value: string;
  /** Worum es geht («Seeblick», «Kühlbox»). */
  title: string;
  /** Ergänzung in klein («Packliste 60 %», «2 laufen morgen ab»). */
  subtitle: string;
  /** Ziel beim Antippen – ein Pfad der App. */
  url: string;
}

/** Fortschritts-Ring, 0–1; null = keiner. */
export interface WidgetProgress {
  ratio: number;
  label: string;
}

export interface WidgetPayload {
  /** Fassung der Nutzlast; das Swift-Teil prüft sie, bevor es zeichnet. */
  version: 2;
  /** Wann gebaut (ISO) – das Widget zeigt es nicht, aber es hilft beim Suchen. */
  builtAt: string;
  trip: WidgetPanel;
  /** Packstand der kommenden Reise; null = keine Liste verknüpft. */
  packing: WidgetProgress | null;
  supplies: WidgetPanel;
  /** Überschrift über der Liste zum Abhaken. */
  tasksTitle: string;
  /** Text, wenn nichts offen ist. */
  tasksEmpty: string;
  /** Ziel beim Antippen NEBEN den Häkchen – die zugehörige Liste. */
  tasksUrl: string;
  /** Die Einträge selbst – höchstens MAX_WIDGET_TASKS. */
  tasks: WidgetTask[];
}

/**
 * Wie viele Zeilen ins mittlere Widget passen, ohne dass es zur Liste
 * wird. Mehr als vier liest im Vorbeigehen ohnehin niemand.
 */
export const MAX_WIDGET_TASKS = 4;

export interface WidgetTripLike {
  id: number;
  title?: string | null;
  location?: string | null;
  spotName?: string | null;
  startDate: string;
  endDate: string;
  /** Verknüpfte Packliste – für den Fortschritts-Ring. */
  packListId?: number | null;
}

export interface WidgetFoodLike {
  expiryDate?: string | null;
}

/** Ein Packlisten-Eintrag, soweit fürs Widget nötig. */
export interface WidgetPackItemLike {
  id: number;
  name: string;
  checked: boolean;
}

/** Eine Ämtli-Zuteilung von heute. */
export interface WidgetChoreLike {
  id: number;
  title: string;
  doneAt: Date | string | null;
}

export interface WidgetInput {
  trips: readonly WidgetTripLike[];
  foodItems: readonly WidgetFoodLike[];
  gearTasks: readonly GearTaskLike[];
  /** Packstand der nächsten Reise, falls eine Liste verknüpft ist. */
  packing: { total: number; checked: number } | null;
  /** Einträge der verknüpften Packliste (vor der Reise). */
  packItems?: readonly WidgetPackItemLike[];
  /** Heutige Ämtli (während der Reise). */
  chores?: readonly WidgetChoreLike[];
  /** Heutiges Datum als ISO-Tag (YYYY-MM-DD). */
  today: string;
  lang: Language;
}

/**
 * Name eines Aufenthalts – aus derselben Quelle wie überall sonst (#329).
 *
 * Vorher stand die Regel hier ein zweites Mal. Das Widget hätte damit
 * einen anderen Namen zeigen können als die App daneben, ohne dass es
 * jemandem auffällt.
 */
const tripName = tripDisplayName;

/**
 * Der laufende oder nächste Aufenthalt.
 *
 * LAUFEND SCHLÄGT KOMMEND: Wer gerade auf dem Platz steht, will nicht
 * lesen, dass in drei Wochen die nächste Reise ansteht.
 */
export function selectWidgetTrip(
  trips: readonly WidgetTripLike[],
  today: string
): { trip: WidgetTripLike; running: boolean } | null {
  const running = trips
    .filter(trip => currentTripDay(trip, today) !== null)
    .sort((a, b) => a.startDate.localeCompare(b.startDate))[0];
  if (running) return { trip: running, running: true };
  const upcoming = trips
    .filter(trip => trip.startDate > today)
    .sort((a, b) => a.startDate.localeCompare(b.startDate))[0];
  return upcoming ? { trip: upcoming, running: false } : null;
}

function tripPanel(input: WidgetInput): WidgetPanel {
  const { lang, today } = input;
  const found = selectWidgetTrip(input.trips, today);
  if (!found) {
    return {
      value: "–",
      title: pick(
        l4(
          "Keine Reise geplant",
          "Aucun voyage prévu",
          "Nessun viaggio",
          "No trip planned"
        ),
        lang
      ),
      subtitle: pick(
        l4(
          "Tippen zum Planen",
          "Toucher pour planifier",
          "Tocca per pianificare",
          "Tap to plan"
        ),
        lang
      ),
      url: "/tagebuch",
    };
  }
  const { trip, running } = found;
  const name = tripName(trip, lang);
  const url = `/tagebuch/${trip.id}`;

  if (running) {
    const day = currentTripDay(trip, today);
    return {
      value: String(day?.day ?? 1),
      title: name,
      subtitle: pick(
        l4(
          `Tag ${day?.day} von ${day?.total}`,
          `Jour ${day?.day} sur ${day?.total}`,
          `Giorno ${day?.day} di ${day?.total}`,
          `Day ${day?.day} of ${day?.total}`
        ),
        lang
      ),
      url,
    };
  }

  // Hier ist `days` immer ≥ 1: Eine Reise, die HEUTE beginnt, ist bereits
  // eine laufende (Anreisetag = Tag 1) und wurde oben behandelt. Ein
  // Zweig für «heute» wäre toter Code – samt vier Übersetzungen, die
  // niemand je zu Gesicht bekäme und die beim nächsten Umbau falsch
  // würden, ohne dass es auffällt.
  const days = daysUntilTrip(trip.startDate, today);
  const subtitle =
    days === 1
      ? pick(
          l4(
            "Anreise morgen",
            "Arrivée demain",
            "Arrivo domani",
            "Arrival tomorrow"
          ),
          lang
        )
      : pick(
          l4(
            `noch ${days} Tage`,
            `encore ${days} jours`,
            `ancora ${days} giorni`,
            `${days} days to go`
          ),
          lang
        );
  return { value: String(days), title: name, subtitle, url };
}

function packingProgress(input: WidgetInput): WidgetProgress | null {
  const packed = input.packing;
  if (!packed || packed.total <= 0) return null;
  const pct = Math.round((packed.checked / packed.total) * 100);
  return {
    ratio: Math.min(1, Math.max(0, packed.checked / packed.total)),
    label: pick(
      l4(
        `Packliste ${pct} %`,
        `Bagages ${pct} %`,
        `Bagagli ${pct} %`,
        `Packing ${pct} %`
      ),
      input.lang
    ),
  };
}

/**
 * Ablaufende Lebensmittel und fällige Pflege – dieselbe Zählung wie beim
 * App-Icon-Zähler (#132), damit die beiden Zahlen nie auseinanderlaufen.
 */
export function countSupplies(input: {
  foodItems: readonly WidgetFoodLike[];
  gearTasks: readonly GearTaskLike[];
  today: string;
}): { expiring: number; due: number } {
  const expiring = input.foodItems.filter(item => {
    const info = expiryInfo(item.expiryDate, input.today);
    return info !== null && info.daysLeft >= 0 && info.daysLeft <= 1;
  }).length;
  const due = input.gearTasks.filter(
    task => gearTaskDue(task, input.today).due
  ).length;
  return { expiring, due };
}

function suppliesPanel(input: WidgetInput): WidgetPanel {
  const { lang } = input;
  const { expiring, due } = countSupplies(input);
  const total = expiring + due;
  if (total === 0) {
    return {
      value: "✓",
      title: pick(
        l4("Alles im Griff", "Tout est en ordre", "Tutto a posto", "All clear"),
        lang
      ),
      subtitle: pick(
        l4(
          "Nichts läuft ab, nichts fällig",
          "Rien n’expire, rien à faire",
          "Nulla scade, nulla in scadenza",
          "Nothing expiring, nothing due"
        ),
        lang
      ),
      url: "/kuehlbox",
    };
  }
  // Der Titel nennt das Dringendere zuerst: Essen verdirbt, Pflege wartet.
  const title =
    expiring > 0
      ? pick(l4("Kühlbox", "Glacière", "Frigo", "Cool box"), lang)
      : pick(l4("Ausrüstung", "Équipement", "Attrezzatura", "Gear"), lang);
  const parts: string[] = [];
  if (expiring > 0) {
    parts.push(
      pick(
        expiring === 1
          ? l4(
              "1 läuft bald ab",
              "1 arrive à expiration",
              "1 sta per scadere",
              "1 expiring soon"
            )
          : l4(
              `${expiring} laufen bald ab`,
              `${expiring} arrivent à expiration`,
              `${expiring} stanno per scadere`,
              `${expiring} expiring soon`
            ),
        lang
      )
    );
  }
  if (due > 0) {
    parts.push(
      pick(
        due === 1
          ? l4(
              "1 Pflege fällig",
              "1 entretien dû",
              "1 manutenzione",
              "1 task due"
            )
          : l4(
              `${due} Pflege fällig`,
              `${due} entretiens dus`,
              `${due} manutenzioni`,
              `${due} tasks due`
            ),
        lang
      )
    );
  }
  return {
    value: String(total),
    title,
    subtitle: parts.join(" · "),
    url: expiring > 0 ? "/kuehlbox" : "/inventar",
  };
}

/**
 * Was im Widget zum Abhaken steht (#327).
 *
 * DIE REGEL FOLGT DEM ORT, AN DEM MAN STEHT: Vor der Reise sind es die
 * nächsten offenen Punkte der verknüpften Packliste – da steht man vor
 * dem Schrank. Während der Reise die heutigen Ämtli – da steht man auf
 * dem Platz. Nach der Reise nichts; dort gibt es keinen Handgriff, den
 * ein Widget abnehmen könnte.
 *
 * OFFENE ZUERST, ERLEDIGTE DAHINTER: Ein Widget mit vier abgehakten
 * Zeilen ist nutzlos. Erledigte bleiben trotzdem sichtbar, solange Platz
 * ist – sonst verschwindet die Zeile im selben Moment, in dem man sie
 * antippt, und man weiss nicht, ob der Tipp angekommen ist.
 */
export function selectWidgetTasks(input: WidgetInput): {
  title: string;
  empty: string;
  url: string;
  tasks: WidgetTask[];
} {
  const { lang } = input;
  const found = selectWidgetTrip(input.trips, input.today);

  if (found?.running) {
    const chores = (input.chores ?? []).map(row => ({
      id: row.id,
      kind: "chore" as const,
      title: row.title,
      checked: row.doneAt !== null,
    }));
    return {
      title: pick(
        l4(
          "Heute im Camp",
          "Aujourd’hui au camp",
          "Oggi al campo",
          "Today at camp"
        ),
        lang
      ),
      empty: pick(
        l4(
          "Keine Ämtli für heute",
          "Aucune tâche aujourd’hui",
          "Nessun turno per oggi",
          "No chores today"
        ),
        lang
      ),
      url: "/aemtli",
      tasks: orderTasks(chores),
    };
  }

  const items = (input.packItems ?? []).map(row => ({
    id: row.id,
    kind: "packing" as const,
    title: row.name,
    checked: row.checked,
  }));
  return {
    title: pick(
      l4(
        "Noch zu packen",
        "Encore à préparer",
        "Ancora da preparare",
        "Still to pack"
      ),
      lang
    ),
    empty: pick(
      l4("Alles gepackt", "Tout est prêt", "Tutto pronto", "All packed"),
      lang
    ),
    url: found?.trip.packListId
      ? `/packlisten/${found.trip.packListId}`
      : "/packlisten",
    tasks: orderTasks(items),
  };
}

/** Offene zuerst, dann erledigte; abgeschnitten auf MAX_WIDGET_TASKS. */
function orderTasks(tasks: readonly WidgetTask[]): WidgetTask[] {
  const open = tasks.filter(task => !task.checked);
  const done = tasks.filter(task => task.checked);
  return [...open, ...done].slice(0, MAX_WIDGET_TASKS);
}

/** Die vollständige Nutzlast bauen. */
export function buildWidgetPayload(input: WidgetInput): WidgetPayload {
  const tasks = selectWidgetTasks(input);
  return {
    version: 2,
    builtAt: `${input.today}T00:00:00Z`,
    trip: tripPanel(input),
    packing: packingProgress(input),
    supplies: suppliesPanel(input),
    tasksTitle: tasks.title,
    tasksEmpty: tasks.empty,
    tasksUrl: tasks.url,
    tasks: tasks.tasks,
  };
}

/**
 * Hat sich etwas geändert, das im Widget sichtbar wäre?
 *
 * Das Neuzeichnen eines Widgets ist beim System ein Budget-Posten – iOS
 * drosselt Apps, die ohne Grund oft nachladen. Verglichen wird deshalb
 * die fertige Nutzlast OHNE `builtAt`: Ein neuer Tag allein ist kein
 * Grund, solange alle Texte gleich bleiben.
 */
export function widgetPayloadChanged(
  previous: WidgetPayload | null,
  next: WidgetPayload
): boolean {
  if (!previous) return true;
  const strip = (p: WidgetPayload) => JSON.stringify({ ...p, builtAt: "" });
  return strip(previous) !== strip(next);
}
